File: autograd/symbol.py
from typing import Any,  List, Optional, Union

import numpy as np

from .ops import Operation

type Inputs = Union[float, int, np.ndarray, list, 'Symbol']
type NonSymbolInputs = Union[float, int, np.ndarray, list]


class Symbol:
    def __init__(
        self,
        value: Optional[NonSymbolInputs] = None,
        operation: Operation = Operation.CONSTANT,
        prev: Optional[List['Symbol']] = None,
        name: Optional[str] = None,
        retain: bool = False,
        requires_grad: Optional[bool] = None,
        **kwargs: Any
    ):
        self.operation = operation
        self.prev = prev if prev is not None else []
        self.name = name
        self.retain = retain
        self.grad: Optional['Symbol'] = None
        self.kwargs = kwargs

        if value is not None:
            self.value = value if isinstance(
                value, np.ndarray) else np.array(value)
        else:
            self.value = None
        
        if self.prev:
            self.requires_grad = any(
                prev_symbol.requires_grad for prev_symbol in self.prev)

            if requires_grad is not None:
                self.requires_grad = requires_grad
        else:
            if requires_grad is not None:
                self.requires_grad = requires_grad
            else:
                self.requires_grad = (operation == Operation.VARIABLE)

    def __repr__(self) -> str:
        name_str = f"{self.name} " if self.name else "Unnamed"
        return f"Symbol(name={name_str}, op={self.operation.name}, requires_grad={self.requires_grad})"

    def _ensure_symbol(self, other: Inputs) -> 'Symbol':
        if not isinstance(other, Symbol):
            return Symbol(value=other, requires_grad=False)
        return other

    def __add__(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.ADD, prev=[self, self._ensure_symbol(other)])

    def __radd__(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.ADD, prev=[self._ensure_symbol(other), self])

    def __mul__(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.MULTIPLY, prev=[self, self._ensure_symbol(other)])

    def __rmul__(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.MULTIPLY, prev=[self._ensure_symbol(other), self])

    def __pow__(self, other: Inputs) -> 'Symbol':

        return Symbol(operation=Operation.POWER, prev=[self, self._ensure_symbol(other)])

    def __matmul__(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.MATMUL, prev=[self, self._ensure_symbol(other)])

    def __rmatmul__(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.MATMUL, prev=[self._ensure_symbol(other), self])

    def __neg__(self) -> 'Symbol':
        return Symbol(operation=Operation.NEG, prev=[self])

    def __sub__(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.SUB, prev=[self, self._ensure_symbol(other)])

    def __rsub__(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.SUB, prev=[self._ensure_symbol(other), self])

    def __truediv__(self, other: Inputs) -> 'Symbol':

        power_node = Symbol(operation=Operation.POWER, prev=[
                            self._ensure_symbol(other), Symbol(value=-1.0)])

        return Symbol(operation=Operation.MULTIPLY, prev=[self, power_node])

    def __rtruediv__(self, other: Inputs) -> 'Symbol':
        power_node = Symbol(operation=Operation.POWER, prev=[
            self, Symbol(value=-1.0)])

        return Symbol(operation=Operation.MULTIPLY, prev=[self._ensure_symbol(other), power_node])

    def exp(self) -> 'Symbol':
        return Symbol(operation=Operation.EXP, prev=[self])

    def log(self) -> 'Symbol':
        return Symbol(operation=Operation.LOG, prev=[self])

    def relu(self) -> 'Symbol':
        return Symbol(operation=Operation.RELU, prev=[self])

    def transpose(self) -> 'Symbol':
        return Symbol(operation=Operation.TRANSPOSE, prev=[self])

    def sum(self, axis: Optional[Union[int, tuple]] = None, keepdims: bool = False) -> 'Symbol':
        return Symbol(operation=Operation.SUM, prev=[self], axis=axis, keepdims=keepdims)

    def mean(self, axis: Optional[Union[int, tuple]] = None, keepdims: bool = False) -> 'Symbol':
        return Symbol(operation=Operation.MEAN, prev=[self], axis=axis, keepdims=keepdims)

    def variance(self, axis: Optional[Union[int, tuple]] = None, keepdims: bool = False) -> 'Symbol':
        return Symbol(operation=Operation.VARIANCE, prev=[self], axis=axis, keepdims=keepdims)

    def softmax(self, axis: Optional[int] = None) -> 'Symbol':
        return Symbol(operation=Operation.SOFTMAX, prev=[self], axis=axis)

    def broadcast_to(self, shape: tuple) -> 'Symbol':
        return Symbol(operation=Operation.BROADCAST_TO, prev=[self], shape=shape)

    def reshape(self, shape: tuple) -> 'Symbol':
        return Symbol(operation=Operation.RESHAPE, prev=[self], shape=shape)

    def expand_dims(self, axis) -> 'Symbol':
        return Symbol(operation=Operation.EXPAND_DIMS, prev=[self], axis=axis)

    def squeeze(self, axis: Optional[int] = None) -> 'Symbol':
        return Symbol(operation=Operation.SQUEEZE, prev=[self], axis=axis)

    def swap_axis(self, axis1: int, axis2: int) -> 'Symbol':
        return Symbol(operation=Operation.SWAP_AXIS, prev=[self], axis1=axis1, axis2=axis2)

    def sqrt(self) -> 'Symbol':
        return Symbol(operation=Operation.SQRT, prev=[self])

    def abs(self) -> 'Symbol':
        return Symbol(operation=Operation.ABS, prev=[self])

    def broadcast_to_match(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.BROADCAST_TO_MATCH, prev=[self, self._ensure_symbol(other)])

    def unbroadcast(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.UNBROADCAST, prev=[self, self._ensure_symbol(other)])

    def reshape_like(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.RESHAPE_LIKE, prev=[self, self._ensure_symbol(other)])

    def size(self, axis=None) -> 'Symbol':
        return Symbol(operation=Operation.SIZE, prev=[self], axis=axis)

    def __gt__(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.GREATER_THAN, prev=[self, self._ensure_symbol(other)])

    def __lt__(self, other: Inputs) -> 'Symbol':
        return Symbol(operation=Operation.LESS_THAN, prev=[self, self._ensure_symbol(other)])

    def sign(self) -> 'Symbol':
        return Symbol(operation=Operation.SIGN, prev=[self]);

    def get_item(self, index: int) -> 'Symbol':
        return Symbol(operation=Operation.GET_ITEM, prev=[self], index=index);

    def scatter_like(self, other: Inputs, index: int) -> 'Symbol':
        return Symbol(operation=Operation.SCATTER_LIKE, prev=[self, self._ensure_symbol(other)], index=index)

File: cpp/src/symbol.cpp
#include "autograd/symbol.hpp"
#include <cassert>
#include <span>

namespace autograd {

// ── Helper
// ────────────────────────────────────────────────────────────────────

static Symbol set_shape(Symbol result, const std::span<int> &shape) {
  uint32_t shape_offset = result.graph->shapes.size();
  for (int s : shape)
    result.graph->shapes.push_back(s);
  result.graph->nodes[result.node_index].shape.offset = shape_offset;
  result.graph->nodes[result.node_index].shape.count =
      static_cast<uint16_t>(shape.size());
  return result;
}

// ── operator[] ───────────────────────────────────────────────────────────────

Symbol Symbol::operator[](int index) const {
  assert(graph != nullptr && "Symbol must be associated with a graph");
  return graph->add_node(Op::GET_ITEM, {node_index}, index);
}

// ── Addition
// ──────────────────────────────────────────────────────────────────

Symbol operator+(Symbol lhs, Symbol rhs) {
  assert(lhs.graph == rhs.graph && "Symbols must belong to the same graph");
  assert(lhs.graph != nullptr && "Symbols must be associated with a graph");
  return lhs.graph->add_node(Op::ADD, {lhs.node_index, rhs.node_index});
}

Symbol operator+(Symbol lhs, float rhs) {
  return lhs + lhs.graph->constant(rhs);
}
Symbol operator+(float lhs, Symbol rhs) { return rhs + lhs; }
Symbol operator+(Symbol lhs, const Eigen::MatrixXf &rhs) {
  return lhs + lhs.graph->constant(rhs);
}
Symbol operator+(const Eigen::MatrixXf &lhs, Symbol rhs) { return rhs + lhs; }

// ── Multiplication
// ────────────────────────────────────────────────────────────

Symbol operator*(Symbol lhs, Symbol rhs) {
  assert(lhs.graph == rhs.graph && "Symbols must belong to the same graph");
  assert(lhs.graph != nullptr && "Symbols must be associated with a graph");
  return lhs.graph->add_node(Op::MULTIPLY, {lhs.node_index, rhs.node_index});
}

Symbol operator*(Symbol lhs, float rhs) {
  return lhs * lhs.graph->constant(rhs);
}
Symbol operator*(float lhs, Symbol rhs) { return rhs * lhs; }
Symbol operator*(Symbol lhs, const Eigen::MatrixXf &rhs) {
  return lhs * lhs.graph->constant(rhs);
}
Symbol operator*(const Eigen::MatrixXf &lhs, Symbol rhs) { return rhs * lhs; }

// ── Negation / Subtraction
// ────────────────────────────────────────────────────

Symbol operator-(Symbol x) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  return x.graph->add_node(Op::NEG, {x.node_index});
}

Symbol operator-(Symbol lhs, Symbol rhs) {
  assert(lhs.graph == rhs.graph && "Symbols must belong to the same graph");
  assert(lhs.graph != nullptr && "Symbols must be associated with a graph");
  return lhs.graph->add_node(Op::SUB, {lhs.node_index, rhs.node_index});
}

Symbol operator-(Symbol lhs, float rhs) {
  return lhs - lhs.graph->constant(rhs);
}
Symbol operator-(float lhs, Symbol rhs) {
  return rhs.graph->constant(lhs) - rhs;
}
Symbol operator-(Symbol lhs, const Eigen::MatrixXf &rhs) {
  return lhs - lhs.graph->constant(rhs);
}
Symbol operator-(const Eigen::MatrixXf &lhs, Symbol rhs) {
  return rhs.graph->constant(lhs) - rhs;
}

// ── Division
// ────────────────────────────────────────────────────────────────── No DIV op
// — mirrors Python: lhs / rhs = lhs * pow(rhs, -1)

Symbol operator/(Symbol lhs, Symbol rhs) {
  assert(lhs.graph == rhs.graph && "Symbols must belong to the same graph");
  assert(lhs.graph != nullptr && "Symbols must be associated with a graph");
  return lhs * pow(rhs, -1.0f);
}

Symbol operator/(Symbol lhs, float rhs) {
  return lhs / lhs.graph->constant(rhs);
}
Symbol operator/(float lhs, Symbol rhs) {
  return rhs.graph->constant(lhs) / rhs;
}
Symbol operator/(Symbol lhs, const Eigen::MatrixXf &rhs) {
  return lhs / lhs.graph->constant(rhs);
}
Symbol operator/(const Eigen::MatrixXf &lhs, Symbol rhs) {
  return rhs.graph->constant(lhs) / rhs;
}

// ── Power
// ─────────────────────────────────────────────────────────────────────

Symbol pow(Symbol base, Symbol exp) {
  assert(base.graph == exp.graph && "Symbols must belong to the same graph");
  assert(base.graph != nullptr && "Symbols must be associated with a graph");
  return base.graph->add_node(Op::POWER, {base.node_index, exp.node_index});
}

Symbol pow(Symbol base, float exp) {
  return pow(base, base.graph->constant(exp));
}
Symbol pow(float base, Symbol exp) {
  return pow(exp.graph->constant(base), exp);
}

// ── Matmul
// ────────────────────────────────────────────────────────────────────

Symbol matmul(Symbol lhs, Symbol rhs) {
  assert(lhs.graph == rhs.graph && "Symbols must belong to the same graph");
  assert(lhs.graph != nullptr && "Symbols must be associated with a graph");
  return lhs.graph->add_node(Op::MATMUL, {lhs.node_index, rhs.node_index});
}

Symbol matmul(Symbol lhs, const Eigen::MatrixXf &rhs) {
  return matmul(lhs, lhs.graph->constant(rhs));
}
Symbol matmul(const Eigen::MatrixXf &lhs, Symbol rhs) {
  return matmul(rhs.graph->constant(lhs), rhs);
}

// ── Unary
// ─────────────────────────────────────────────────────────────────────

Symbol exp(Symbol x) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  return x.graph->add_node(Op::EXP, {x.node_index});
}
Symbol log(Symbol x) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  return x.graph->add_node(Op::LOG, {x.node_index});
}
Symbol relu(Symbol x) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  return x.graph->add_node(Op::RELU, {x.node_index});
}
Symbol sqrt(Symbol x) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  return x.graph->add_node(Op::SQRT, {x.node_index});
}
Symbol abs(Symbol x) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  return x.graph->add_node(Op::ABS, {x.node_index});
}
Symbol sign(Symbol x) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  return x.graph->add_node(Op::SIGN, {x.node_index});
}
Symbol transpose(Symbol x) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  return x.graph->add_node(Op::TRANSPOSE, {x.node_index});
}

// ── Reductions
// ────────────────────────────────────────────────────────────────

Symbol sum(Symbol x, std::optional<int> axis, bool keepdims) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  int32_t ax = axis.value_or(-1);
  return x.graph->add_node(Op::SUM, {x.node_index}, ax, UINT32_MAX, keepdims);
}
Symbol mean(Symbol x, std::optional<int> axis, bool keepdims) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  int32_t ax = axis.value_or(-1);
  return x.graph->add_node(Op::MEAN, {x.node_index}, ax, UINT32_MAX, keepdims);
}
Symbol variance(Symbol x, std::optional<int> axis, bool keepdims) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  int32_t ax = axis.value_or(-1);
  return x.graph->add_node(Op::VARIANCE, {x.node_index}, ax, UINT32_MAX,
                           keepdims);
}
Symbol softmax(Symbol x, std::optional<int> axis) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  int32_t ax = axis.value_or(-1);
  return x.graph->add_node(Op::SOFTMAX, {x.node_index}, ax);
}
Symbol size(Symbol x, std::optional<int> axis) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  int32_t ax = axis.value_or(-1);
  return x.graph->add_node(Op::SIZE, {x.node_index}, ax);
}

// ── Shape
// ─────────────────────────────────────────────────────────────────────

Symbol broadcast_to(Symbol x, std::span<int> shape) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  Symbol result = x.graph->add_node(Op::BROADCAST_TO, {x.node_index});
  return set_shape(result, shape);
}
Symbol broadcast_to_match(Symbol x, Symbol other) {
  assert(x.graph == other.graph && "Symbols must belong to the same graph");
  assert(x.graph != nullptr && "Symbols must be associated with a graph");
  return x.graph->add_node(Op::BROADCAST_TO_MATCH,
                           {x.node_index, other.node_index});
}
Symbol reshape(Symbol x, std::span<int> shape) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  Symbol result = x.graph->add_node(Op::RESHAPE, {x.node_index});
  return set_shape(result, shape);
}
Symbol reshape_like(Symbol x, Symbol other) {
  assert(x.graph == other.graph && "Symbols must belong to the same graph");
  assert(x.graph != nullptr && "Symbols must be associated with a graph");
  return x.graph->add_node(Op::RESHAPE_LIKE, {x.node_index, other.node_index});
}
Symbol expand_dims(Symbol x, int axis) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  return x.graph->add_node(Op::EXPAND_DIMS, {x.node_index}, axis);
}
Symbol squeeze(Symbol x, std::optional<int> axis) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  int32_t ax = axis.value_or(-1);
  return x.graph->add_node(Op::SQUEEZE, {x.node_index}, ax);
}
Symbol swap_axis(Symbol x, int axis1, int axis2) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  return x.graph->add_node(Op::SWAP_AXIS, {x.node_index}, axis1, axis2);
}
Symbol unbroadcast(Symbol x, Symbol other) {
  assert(x.graph == other.graph && "Symbols must belong to the same graph");
  assert(x.graph != nullptr && "Symbols must be associated with a graph");
  return x.graph->add_node(Op::UNBROADCAST, {x.node_index, other.node_index});
}

// ── Comparison
// ────────────────────────────────────────────────────────────────

Symbol operator==(Symbol lhs, Symbol rhs) {
  assert(lhs.graph == rhs.graph && "Symbols must belong to the same graph");
  assert(lhs.graph != nullptr && "Symbols must be associated with a graph");
  return lhs.graph->add_node(Op::EQUALS_TO, {lhs.node_index, rhs.node_index});
}
Symbol operator==(Symbol lhs, float rhs) {
  return lhs == lhs.graph->constant(rhs);
}
Symbol operator==(float lhs, Symbol rhs) { return rhs == lhs; }

Symbol operator>(Symbol lhs, Symbol rhs) {
  assert(lhs.graph == rhs.graph && "Symbols must belong to the same graph");
  assert(lhs.graph != nullptr && "Symbols must be associated with a graph");
  return lhs.graph->add_node(Op::GREATER_THAN,
                             {lhs.node_index, rhs.node_index});
}
Symbol operator>(Symbol lhs, float rhs) {
  return lhs > lhs.graph->constant(rhs);
}
Symbol operator>(float lhs, Symbol rhs) {
  return rhs.graph->constant(lhs) > rhs;
}

Symbol operator>=(Symbol lhs, Symbol rhs) {
  assert(lhs.graph == rhs.graph && "Symbols must belong to the same graph");
  assert(lhs.graph != nullptr && "Symbols must be associated with a graph");
  return lhs.graph->add_node(Op::GREATER_THAN_OR_EQUAL,
                             {lhs.node_index, rhs.node_index});
}
Symbol operator>=(Symbol lhs, float rhs) {
  return lhs >= lhs.graph->constant(rhs);
}
Symbol operator>=(float lhs, Symbol rhs) {
  return rhs.graph->constant(lhs) >= rhs;
}

Symbol operator<(Symbol lhs, Symbol rhs) {
  assert(lhs.graph == rhs.graph && "Symbols must belong to the same graph");
  assert(lhs.graph != nullptr && "Symbols must be associated with a graph");
  return lhs.graph->add_node(Op::LESS_THAN, {lhs.node_index, rhs.node_index});
}
Symbol operator<(Symbol lhs, float rhs) {
  return lhs < lhs.graph->constant(rhs);
}
Symbol operator<(float lhs, Symbol rhs) {
  return rhs.graph->constant(lhs) < rhs;
}

Symbol operator<=(Symbol lhs, Symbol rhs) {
  assert(lhs.graph == rhs.graph && "Symbols must belong to the same graph");
  assert(lhs.graph != nullptr && "Symbols must be associated with a graph");
  return lhs.graph->add_node(Op::LESS_THAN_OR_EQUAL,
                             {lhs.node_index, rhs.node_index});
}
Symbol operator<=(Symbol lhs, float rhs) {
  return lhs <= lhs.graph->constant(rhs);
}
Symbol operator<=(float lhs, Symbol rhs) {
  return rhs.graph->constant(lhs) <= rhs;
}

// ── Indexing
// ──────────────────────────────────────────────────────────────────

Symbol get_item(Symbol x, int index) {
  assert(x.graph != nullptr && "Symbol must be associated with a graph");
  return x.graph->add_node(Op::GET_ITEM, {x.node_index},
                           static_cast<uint32_t>(index));
}
Symbol scatter_like(Symbol x, Symbol other, int index) {
  assert(x.graph == other.graph && "Symbols must belong to the same graph");
  assert(x.graph != nullptr && "Symbols must be associated with a graph");
  return x.graph->add_node(Op::SCATTER_LIKE, {x.node_index, other.node_index},
                           static_cast<uint32_t>(index));
}

} // namespace autograd
